pbsnodes -r does not clear mom state, offline_vnodes offline all vnodes

arungrover opened the pull request:
<!--- Please review your changes in preview mode -->
<!--- Provide a general summary of your changes in the Title above -->

#### Bug/feature Description
There are two problems we are dealing with here - 
   - when a mom hook fail action is set to offline_vnodes it offlines all the vnodes associating with that mom. It does not check whether or not the same vnodes are being reported by other moms as well and there would be a way to use these vnodes using other moms. 
   - When a mom has been offlined and admin issues 'pbsnodes -r' command for that mom host, pbs server resets the state of the node back to free but it does not clear the mom state internally which means that any job that ends up on that mom will not run.

To reproduce the problems mentioned above perform following steps
 - Have two moms having the same set of vnode definitions
 - Create a mom execjob_begin hook that sleeps for a long period of time
 - Set the fail action for this hook as 'offline_vnodes'
 - Set hook alarm smaller than the sleep time in the hook
 - Submit a job that runs on one of the moms
 - On a buggy build you will see the mom and all the vnodes being turned offline when the execjob_begin hook is alarmed out

Now to reproduce pbsnode -r problem do all the steps mentioned above and 
 - issue pbsnodes -r for the mom host that has been marked as offline
 - Submit another job which will run on the vnode that was previously marked as offline
 - On a buggy build you will notice that job stays in queued state even when pbsnodes is reporting it as free


#### Affected Platform(s)
* All platforms

#### Analysis / Design
*  [External design for offline_vnode problem](https://pbspro.atlassian.net/wiki/spaces/PD/pages/195002450/offline+vnodes+should+only+offline+vnodes+belonging+to+more+than+one+Mom+when+all+the+MoMs+are+offline)
* [pbspro community forum discussion](http://community.pbspro.org/t/offline-vnodes-should-only-offline-vnodes-belonging-to-more-than-one-mom-when-all-the-moms-are-offline/864)


#### Solution Description
* Fix includes check to make sure that when vnodes are offlined it checks that all the moms are also marked offline. Fix also includes a change where while setting node state it also sets its mom state internally.

#### Checklist:
<!--- Use the preview button to see the checkboxes/links properly. -->
<!--- Go over the following points, and put an `x` (without spaces around it) in the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [x] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
   - [x] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/spaces/PD/pages/195002450/Draft+offline+vnodes+should+only+offline+vnodes+belonging+to+more+than+one+Mom+when+all+the+MoMs+are+offline)**)

- [x] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
   - [ ] I have added  **manual test(s) to this pull request and explained why PTL is not appropriate** for this case.
- [x] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [x] I have attached **test logs to this pull request** as evidence of testing/verification.


__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__


[ptl_dev_test.txt](https://github.com/PBSPro/pbspro/files/1768592/ptl_dev_test.txt)

bhroam commented on test/tests/functional/pbs_offline_vnodes.py:
```diff
@@ -0,0 +1,211 @@
+            # free
+            for v1 in vlist:
+                if first_vnode == 1:
+                    first_vnode += 1
```
Since we're only setting vname once, why don't you break?

arungrover replied:
Well after that I need to check the state of all vnodes reported so I can not break out of the loop

bhroam commented on test/tests/functional/pbs_offline_vnodes.py:
```diff
@@ -0,0 +1,211 @@
+        self.server.manager(MGR_CMD_SET, HOOK, {'enabled': 'False'}, id="h1")
+        self.server.manager(MGR_CMD_SET, SCHED, {'scheduling': 'True'})
+        self.server.expect(JOB, {ATTR_state: 'R'}, id=jid2)
+        self.server.delete(jid2, wait=True)
```
No need to delete the job, it'll happen in revert_to_defaults()

arungrover replied:
done

arungrover commented on test/tests/functional/pbs_offline_vnodes.py:
```diff
@@ -0,0 +1,211 @@
+        TestFunctional.setUp(self)
+
+    def tearDown(self):
+        self.server.manager(MGR_CMD_DELETE, NODE, id='@default')
```
I've removed it. Thanks for catching it

arungrover commented on test/tests/functional/pbs_offline_vnodes.py:
```diff
@@ -0,0 +1,211 @@
+            self.server.manager(MGR_CMD_DELETE, NODE,
+                                id=self.moms.values()[1].shortname)
+        except PbsManagerError as e:
+            self.assertTrue("Unknown node" in e.msg[0])
```
I've removed this as well

arungrover commented on test/tests/functional/pbs_offline_vnodes.py:
```diff
@@ -0,0 +1,211 @@
+                           id=self.moms.values()[0].shortname)
+        j1.set_sleep_time(1000)
+        jid = self.server.submit(j1)
+        self.server.manager(MGR_CMD_SET, SERVER, {'scheduling': '0'})
```
Done

arungrover commented on test/tests/functional/pbs_offline_vnodes.py:
```diff
@@ -0,0 +1,211 @@
+
+    def test_multi_mom_hook_failure_affects_vnode(self):
+        """
+        Run an execjob_begin hook that sleep for sometime,
```
Done

arungrover commented:
[dev_ptl_test2.txt](https://github.com/PBSPro/pbspro/files/1769523/dev_ptl_test2.txt)

vccardenas commented on src/server/node_manager.c:
```diff
@@ -467,8 +467,9 @@ get_addr_of_nodebyname(char *name, unsigned int *port)
-	Set_ALL_State_All_Down,	 /* set set on vnodes when all Moms are down */
-	Set_All_State_Regardless /* set set on vnodes regardless */
+	Set_ALL_State_All_Down,	  /* set set on vnodes when all Moms are down */
+	Set_All_State_Regardless, /* set set on vnodes regardless */
```
set appears twice in the previous two lines

arungrover replied:
I'll fix that

vccardenas commented on test/tests/functional/pbs_offline_vnodes.py:
```diff
@@ -0,0 +1,204 @@
+        In case of a single mom reporting vnodes, it should mark
+        all the vnodes and mom as offline.
+        """
+        self.create_multi_vnodes(1, True)
```
In the case of Cray, you start out with two login nodes. You might want to disable one of them so that your test will truly have one mom.

vccardenas replied:
This single mom test might be good to run on the Altix/UV. But the setup requires two moms.

vccardenas commented on test/tests/functional/pbs_offline_vnodes.py:
```diff
@@ -0,0 +1,204 @@
+                           id=self.moms.values()[0].shortname)
+        j1.set_sleep_time(1000)
+        jid = self.server.submit(j1)
+        self.server.manager(MGR_CMD_SET, SERVER, {'scheduling': 'False'})
```
If you only have one login/mom then you might not need to turn off scheduling here.

lisa-altair commented on test/tests/functional/pbs_offline_vnodes.py:
```diff
@@ -0,0 +1,204 @@
+            NODE, {ATTR_NODE_state: 'offline'},
+            id=self.moms.values()[0].shortname)
+        self.server.expect(NODE, {ATTR_NODE_state: 'free'},
+                           id=self.moms.values()[1].shortname)
```
Sometimes when you check for the node/vnode state you set max_attempts and an interval, other times you don't.  Is there a reason?  I think you should check a few times in all places, just in case.
Here it might be because you don't expect the state of mom1 to change...but just in case I think the test should wait a little to see if it does change.

arungrover replied:
Ideally, it should not take more than a second or so to update the state that is why added max_attempts and interval in a few places. I'll update all other instances.

lisa-altair commented on test/tests/functional/pbs_offline_vnodes.py:
```diff
@@ -0,0 +1,204 @@
+
+        else:
+            self.server.expect(NODE, {ATTR_NODE_state: 'free'}, id='vnode[0]')
+            self.server.expect(NODE, {ATTR_NODE_state: 'free'}, id='vnode[1]')
```
Add max_attempts and an interval here too.

lisa-altair commented on test/tests/functional/pbs_offline_vnodes.py:
```diff
@@ -0,0 +1,204 @@
+        jid2 = self.server.submit(j2)
+        self.du.run_cmd(self.server.hostname, cmd=pbsnodes_reset, sudo=True)
+        self.server.expect(NODE, {ATTR_NODE_state: 'free'},
+                           id=self.moms.values()[0].shortname)
```
Add max_attempts and an interval here too.

arungrover replied:
Done

lisa-altair commented on test/tests/functional/pbs_offline_vnodes.py:
```diff
@@ -0,0 +1,204 @@
+        self.server.expect(NODE, {ATTR_NODE_state: 'free'},
+                           id=self.moms.values()[0].shortname)
+        self.server.expect(NODE, {ATTR_NODE_state: 'free'},
+                           id=self.moms.values()[1].shortname)
```
Should max_attempts and interval be set on these 2 expect() too?

arungrover replied:
Done

lisa-altair commented on test/tests/functional/pbs_offline_vnodes.py:
```diff
@@ -0,0 +1,204 @@
+        self.server.expect(NODE, {ATTR_NODE_state: 'free'},
+                           id=self.moms.values()[0].shortname)
+        self.server.manager(MGR_CMD_SET, SCHED, {'scheduling': 'True'})
+        self.server.expect(JOB, {ATTR_state: 'R'}, id=jid2)
```
Since there is a second mom that has never been marked offline, I believe this test isn't really proving that the pbnodes -r problem is fixed because even with the bug the job can be run from the second mom.  Maybe you should either cause the second mom to be marked offline too.  Or add this part of the test to the single mom test above.

arungrover replied:
That makes sense. Thanks for catching this. I'll move "pbsnodes -r" test to single node test.

arungrover commented on test/tests/functional/pbs_offline_vnodes.py:
```diff
@@ -0,0 +1,204 @@
+        self.create_multi_vnodes(2, False)
+        self.create_mom_hook()
+
+        # set one  natural node to have higher ncpus than the other one
```
Your understanding is correct. I'll add a comment.

arungrover commented:
Attaching t
[dev_ptl_test2.txt](https://github.com/PBSPro/pbspro/files/1790244/dev_ptl_test2.txt)
est logs

lisa-altair commented on test/tests/functional/pbs_offline_vnodes.py:
```diff
@@ -86,39 +87,80 @@ def test_single_mom_hook_failure_affects_vnode(self):
+        self.server.expect(NODE, {ATTR_NODE_state: 'free'},
+                           id=self.moms.values()[0].shortname, max_attempts=3,
+                           interval=2)
+        self.server.manager(MGR_CMD_SET, SCHED, {'scheduling': 'True'})
```
Since you don't make scheduling false anymore, you probably don't need to set it to true.  Submitting the job should cause a scheduling cycle to run.

arungrover replied:
Yes this step is not needed.

lisa-altair commented on test/tests/functional/pbs_offline_vnodes.py:
```diff
@@ -130,10 +172,11 @@ def test_multi_mom_hook_failure_affects_vnode(self):
         self.create_mom_hook()
 
-        # set one  natural node to have higher ncpus than the other one
+        # set one  natural node to have higher ncpus than the other one so
```
nit: extra space before natural

arungrover replied:
I'll change it

lisa-altair commented on test/tests/functional/pbs_offline_vnodes.py:
```diff
@@ -86,39 +87,80 @@ def test_single_mom_hook_failure_affects_vnode(self):
+        mom_host = self.moms.values()[0].shortname
+        pbs_exec = self.server.pbs_conf['PBS_EXEC']
+        pbsnodes_cmd = os.path.join(pbs_exec, 'bin', 'pbsnodes')
+        pbsnodes_reset = pbsnodes_cmd + ' -r ' + mom_host
```
Shouldn't we also check that all of the nodes are free (i.e. no longer offline)?

arungrover replied:
I think that is not needed because there wasn't a problem with "pbsnodes -r" not freeing up the nodes.

lisa-altair replied:
Okay.  Thanks.

lisa-altair commented on test/tests/functional/pbs_offline_vnodes.py:
```diff
@@ -162,43 +205,25 @@ def test_multi_mom_hook_failure_affects_vnode(self):
+                               max_attempts=3, interval=2)
+            self.server.expect(NODE, {ATTR_NODE_state: 'free'}, id='vnode[1]',
+                               max_attempts=3, interval=2)
+        self.server.manager(MGR_CMD_DELETE, NODE, id="@default")
```
Out of curiosity, why do we have to delete the nodes?  Won't the next setUp do that anyway?

arungrover replied:
You are right, this is not needed.

lisa-altair commented on test/tests/functional/pbs_offline_vnodes.py:
```diff
@@ -0,0 +1,229 @@
+
+    def setUp(self):
+        if len(self.moms.values()) != 2:
+            self.skipTest("Provide 2 moms while invoking test")
```
Hey @arungrover, since the first test is really a single mom test, can we move this out of here and into the multi-mom test case?  That way the single mom test can get run most of the time.

arungrover replied:
I'll change it. Thanks!

arungrover commented on test/tests/functional/pbs_offline_vnodes.py:
```diff
@@ -86,39 +87,80 @@ def test_single_mom_hook_failure_affects_vnode(self):
+
+        # Set mom sync hook timeout to be a low value because if mom fails to
+        # get the hook after disabling it then next sync will happen after
+        # 2 minutes by default.
```
I'll add it.

arungrover commented:
[test_on_linux_single_mom.txt](https://github.com/PBSPro/pbspro/files/1804920/test_on_linux_single_mom.txt)
[test_on_linux.txt](https://github.com/PBSPro/pbspro/files/1804921/test_on_linux.txt)
[test_on_UV.txt](https://github.com/PBSPro/pbspro/files/1804922/test_on_UV.txt)

lisa-altair commented on test/tests/functional/pbs_offline_vnodes.py:
```diff
@@ -90,26 +83,28 @@ def test_single_mom_hook_failure_affects_vnode(self):
         self.server.expect(
             NODE, {ATTR_NODE_state: 'offline'},
-            id=self.moms.values()[0].shortname, max_attempts=3, interval=2)
+            id=single_mom.shortname, max_attempts=20, interval=2)
```
Earlier you had expected the check to be quick and that's why max_attempts=3.  But now it is set to 20.  Is that because you added UV testing?

arungrover replied:
Yes Lisa this is because of UV testing. I found out that it took longer on a UV to set the nodes to offline state. Not sure why it was so but thought it is better to increase the number of attempts.

lisa-altair replied:
Okay.  Thanks for confirming.

lisa-altair commented on test/tests/functional/pbs_offline_vnodes.py:
```diff
@@ -90,26 +83,28 @@ def test_single_mom_hook_failure_affects_vnode(self):
+            self.server.manager(MGR_CMD_DELETE, NODE, id="@default",
+                                expect=True)
+        single_mom = self.moms.values()[0]
+
```
Do we need to add the following code in case the mom provided has an unwanted vnodedef file (sorry I can't get the formatting of the comment right)
```
if self.moms.values()[0].has_vnode_defs():
       self.moms.values()[0].delete_vnode_defs()
```

arungrover replied:
I think so, this is just to safe guard that the two moms being added do not come with already existing vnode definitions.

lisa-altair replied:
I meant do we need to add that code to the single mom test case.  :)

arungrover replied:
oh ok, I get it. I can add it, but if you are okay I'd like not to touch the single node test. The test is very delicate and is now working on all 3 platforms (cray, uv and normal linux). I'm worried that changing it may cause trouble on one or more of these platforms.

lisa-altair replied:
Since it's working then it seems the change is not necessary.  I certainly don't want to upset the delicate balance if it is working.  :)  Thanks for the response.

arungrover commented:
[test_on_linux_single.txt](https://github.com/PBSPro/pbspro/files/1808700/test_on_linux_single.txt)
[test_on_linux.txt](https://github.com/PBSPro/pbspro/files/1808701/test_on_linux.txt)
[test_on_UV.txt](https://github.com/PBSPro/pbspro/files/1808702/test_on_UV.txt)

agrawalravi90 commented on test/tests/functional/pbs_offline_vnodes.py:
```diff
@@ -0,0 +1,262 @@
+        self.create_mom_hook()
+
+        j1 = Job(TEST_USER)
+        self.server.expect(NODE, {ATTR_NODE_state: 'free'},
```
also, you might not need this expect if you set expect=True in create_vnodes() inside create_multi_vnodes()

agrawalravi90 commented on test/tests/functional/pbs_offline_vnodes.py:
```diff
@@ -0,0 +1,262 @@
+                           interval=2)
+        j1.set_sleep_time(1000)
+        jid = self.server.submit(j1)
+        self.server.expect(JOB, {ATTR_state: 'Q'}, id=jid)
```
Can you please add a comment here explaining why you expect the job to be in 'Q' state and the node to be offline? I know you explained it in the docstring, but code comments help more.

agrawalravi90 commented on test/tests/functional/pbs_offline_vnodes.py:
```diff
@@ -0,0 +1,262 @@
+        vname = "vnode[0]"
+        if self.is_cray is True:
+            vlist = []
+            first_vnode = 1
```
how about setting vname to None when you enter the if block and just checking for vname inside the "for v1 in vlist" loop? if it's None, set it to v1. That way you won't need 'first_vnode' at all.

agrawalravi90 replied:
in fact, you don't really need to set vname before the if-else blocks, you could just check whether vname is None after the if-else blocks, if it is then set it to 'vnode[0]'

agrawalravi90 commented on test/tests/functional/pbs_offline_vnodes.py:
```diff
@@ -0,0 +1,262 @@
+                    VNODE, {'state': 'offline'}, id=v1, max_attempts=3,
+                    interval=2)
+        elif cpuset_mom is True:
+            first_vnode = 1
```
same comment here about removing first_vnode

agrawalravi90 commented on test/tests/functional/pbs_offline_vnodes.py:
```diff
@@ -0,0 +1,262 @@
+                        id=v1['resources_available.vnode'], max_attempts=3,
+                        interval=2)
+        else:
+            self.server.expect(
```
didn't you create just one vnode? how can there be 2?

agrawalravi90 commented on test/tests/functional/pbs_offline_vnodes.py:
```diff
@@ -0,0 +1,262 @@
+        # 2 minutes by default and we don't want to wait that long.
+        self.server.manager(MGR_CMD_SET, SERVER,
+                            {'sync_mom_hookfiles_timeout': '5'})
+        self.server.manager(MGR_CMD_SET, HOOK, {'enabled': 'False'}, id="h1")
```
Can you please change "create_mom_hook()" to return the id of the hook that it created, store that in a variable when calling the function and use that variable here instead of "h1"?

arungrover replied:
Since both these tests need to create only one hook I just made it a function for simplicity. If you feel strongly about changing it then I can do that otherwise I'd like to keep it as is.

agrawalravi90 commented on test/tests/functional/pbs_offline_vnodes.py:
```diff
@@ -0,0 +1,262 @@
+                            id=self.moms.values()[1].shortname)
+
+        j1 = Job(TEST_USER)
+        self.server.expect(NODE, {ATTR_NODE_state: 'free'},
```
If you say expect=True in create_vnodes() inside create_multi_vnodes() then you probably won't need these expects here.

agrawalravi90 commented on test/tests/functional/pbs_offline_vnodes.py:
```diff
@@ -0,0 +1,262 @@
+        self.create_multi_vnodes(2)
+        self.create_mom_hook()
+
+        # set one natural node to have higher ncpus than the other one so
```
just a suggestion: why not just have the job ask for 'vnode' to be one of the natural vnodes' resources_available.vnode value? that's a sure way of making the job run on a particular vnode.

arungrover replied:
Well I can do that but even with ncpus it is almost doing the same thing. So if you do not feel strongly about it, I'd like to retain it as is.

arungrover commented:
@agrawalravi90 I've addressed your comments. Please have a look.
@lisa-altair there has been some more changes to test script (mostly refactoring) please have a look at it again.

arungrover commented:
[test_on_linux_single.txt](https://github.com/PBSPro/pbspro/files/1817501/test_on_linux_single.txt)
[test_on_linux.txt](https://github.com/PBSPro/pbspro/files/1817502/test_on_linux.txt)
[test_on_UV.txt](https://github.com/PBSPro/pbspro/files/1817503/test_on_UV.txt)

vccardenas commented:
@arungrover , the changes look good to me.
